### PR TITLE
(SIMP-4119) Create gitlab_user fact

### PR DESCRIPTION
**Pull request by lnemsick-simp**
Created gitlab_user fact to indicate whether the local user used
for gitlab-shell exists.

SIMP-4119 #comment create gitlab_user fact

### CODE REVIEW

File: lib/facter/gitlab_user.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+# Return the username of the gitlab-shell user
+#
+# Returns nil if
+# * /var/opt/gitlab/gitlab-shell/config.yml does not exist
+# * /var/opt/gitlab/gitlab-shell/config.yml is malformed YAML
+# * 'user' is not specified in /var/opt/gitlab/gitlab-shell/config.yml
+# * The user specified in /var/opt/gitlab/gitlab-shell/config.yml does
+#   not exist in /etc/passwd.
+#
+Facter.add(:gitlab_user) do
```
Comment (jeefberkey) [inline lib/facter/gitlab_user.rb]: Is the gitlab-shell user different than the user the gitlab service runs as? The name could be more specific

File: lib/facter/gitlab_user.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+  setcode do
+    require 'yaml'
+    gitlab_shell_user = nil
+    begin
+      gitlab_shell_yaml = YAML.load_file(gitlab_shell_conf)
+      # YAML.load_file can return nil or raise Psych::SyntaxError
+      # when YAML is malformed
+      if gitlab_shell_yaml and gitlab_shell_yaml['user']
+        local_users = File.read('/etc/passwd')
```
Comment (trevor-vaughan) [inline lib/facter/gitlab_user.rb]: This should use the `Etc` ruby library for best portability

File: lib/facter/gitlab_user.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        unless local_users.match(%r{^#{gitlab_shell_yaml['user']}:}).nil?
+          gitlab_shell_user = gitlab_shell_yaml['user']
+        end
+      end
+    rescue Psych::SyntaxError
```
Comment (trevor-vaughan) [inline lib/facter/gitlab_user.rb]: We might just want to rescue `StandardError` here since if anything blows up we really don't want to fail.

Comment (lnemsick-simp) [reply lib/facter/gitlab_user.rb]: @trevor-vaughan I'm not a big fan of blanket rescues, as that can hide real problems.  I would rather understand the libraries I am using and catch the appropriate exceptions.

File: lib/facter/gitlab_user.rb (hunk 1 above)
Comment (jeefberkey) [reply lib/facter/gitlab_user.rb]: The fact follows gitlab's naming conventions, however confusing I find them. Best to have it follow what's expected then change it here.

:+1:

File: lib/facter/gitlab_user.rb [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+      # when YAML is malformed
+      if gitlab_shell_yaml and gitlab_shell_yaml['user']
+        begin
+          Etc.getpwnam(gitlab_shell_yaml['user'])
+          gitlab_user = gitlab_shell_yaml['user']
+        rescue ArgumentError
+          # user does not exist
+        end
+      end
+    rescue Psych::SyntaxError
```
Comment (trevor-vaughan) [inline lib/facter/gitlab_user.rb]: If this isn't going to be made more generic, then the `confine` above should be more specific that this needs to be a 'readable file' and not a directory or something else.

**Pull request merged**